Add 'waitFor' and deprecate 'await'

theblixguy opened the pull request:
This PR deprecates the `await` API and renames it to `waitFor`. This is because the `await` keyword is now reserved for the new concurrency features in Swift 5.5 and calling it freely will lead to compile errors.

These changes are done in a fully backwards-compatible way so older projects can continue using the `await` API.

Fixes https://github.com/theappbusiness/TABTestKit/issues/151
---

#### Pre-merge checklist

Before merging any PR, please check the following common things that should be done beforehand. These aren't all always required, so just check the box if it doesn't apply.

- [X] **When adding files, make sure they're added to the right target**. If you're adding new files that should be bundled up with Cocoapods etc, they need to be added to the `TABTestKit` target, not `Pods-TABTestKit_Example` etc.
- [X] **Run `pod install` to ensure that the latest changes are in the Example project**. Without this, Carthage might not see the latest changes.
- [X] **Added and updated tests where possible**. This isn't always possible but try wherever you can. The example app contains UI tests to test many of the TABTestKit features.
- [X] **Updated the `CHANGELOG`**. For any changes pending a release, add to the Pending section. For releases, move everything pending to the release section.
- [X] **Updated the `README`**. Add info for any new features, update existing info for anything that's changed or needs extra info.

theblixguy commented:
Need to fix the formatting!

zacoid55 commented on TABTestKit/Classes/Contexts/InteractionContext.swift:
```diff
@@ -35,11 +35,19 @@ public extension InteractionContext {
+    #if swift(>=5.5)
+      states.forEach { element.waitFor($0) }
+    #else
+		  states.forEach { element.await($0) }
```
Has the indentation gone a bit mad on this one and the one below?

theblixguy replied:
Yeah I need to fix the indentation. I'll do that soon

Neil3079 commented:
Yo @theblixguy any update on this thread? iOS 15 is getting pretty close 😬 Need any help on this?

theblixguy commented:
@Neil3079 just need to fix some formatting here (and make sure tests pass, seems like some are failing?). Will try to do that shortly!

tl-riccardo-Cipolleschi commented on TABTestKit/Classes/Protocols/Element.swift:
```diff
@@ -68,27 +68,60 @@ public extension Element {
+  func await(not states: ElementAttributes.State..., timeout: TimeInterval = 30) {
+    guard !states.isEmpty else { XCTFatalFail("You must provide at least one state!") }
+    states.forEach { state in
+      XCTAssertTrue(determine(state, timeout: timeout), "Failed awaiting element to be \(state) with timeout \(timeout)")
```
```suggestion
      XCTAssertTrue(determine(not: state, timeout: timeout), "Failed awaiting element to be \(state) with timeout \(timeout)")
```
There is a bug here. We are implementing `await(not:` and we are checking `determine(state` instead of `determine(not: state`.

Neil3079 replied:
@theblixguy are you ok if we hit commit suggestion here and see if the tests pass?

theblixguy replied:
Sure

tl-riccardo-Cipolleschi reviewed (changes requested):
There is a bug in the `await(not:` function. I guess that it's the reason why the tests are not passing.

Neil3079 commented:
It looks like the tests are failing because they can't select the right Xcode versions. It looks like more of an issue with the workflows themselves, maybe GitHub no longer support Xcode 10 etc?

theblixguy commented:
> It looks like the tests are failing because they can't select the right Xcode versions. It looks like more of an issue with the workflows themselves, maybe GitHub no longer support Xcode 10 etc?

Yeah, I could drop them and bump some versions up (I saw an error about Xcode 12.1 not being available and I can bump it to 12.5.1).

theblixguy commented:
Looks like Xcode 10.3 doesn't exist on macOS 11 (which is now used for `macos-latest`) hence why, I changed it to macOS 10.15 for the Xcode 10.3 workflow job. Also updated to use Xcode 12.5.1 and iOS 14.5 in the workflow jobs using macOS 11. Let's see if everything passes now!

theblixguy commented:
Tests seem a bit flaky, I don't have Xcode 12.5.1 but I have 12.5 and I tested locally on iPhone 12/iOS 14.5 sim and all the tests passed. IDK why they fail on CI 🤔 I've tweaked the yml file to use 12.5 instead of 12.5.1, let's see if that makes any difference.

theblixguy commented:
Oh wow all tests pass now ✨ Just need to fix the indentation now.

theblixguy commented:
I am actually finding it a bit difficult to fix the indentation here (I tried a few combinations of settings but it still doesn't result in the right output). What I am thinking is to merge this as-is (code still looks fine in Xcode, it's just here that it looks a bit off) and then create a new indentation PR to update all the files (seems like it's a mix of different ones due to people different people working on it) and maybe use https://github.com/creyD/prettier_action to auto-fix indentations on PRs.

Neil3079 commented:
@theblixguy I think that makes sense, no point in blocking this release for indentation but would be good to get it sorted properly
